rfcs/56-better-reflect.md

```diff
@@ -0,0 +1,310 @@
+only `Reflect`.
+
+## User-facing explanation
+
+`Reflect` let you navigate any type regardless independently of their shape
+with the `reflect_ref` and `reflect_mut` methods. `CanonReflect` is a special
+case of `Reflect` that also let you convert into a concrete type using the
+`Any` methods. To get a `CanonReflect` from a `Reflect`, you should use any
+of the `make_canon` methods on `dyn Reflect`, `TypeRegistry` or
+`TypeRegistration`.
+
+Any type derived with `#[derive(Reflect)]` implements `CanonReflect`. The only
+types that implements `Reflect` but not `CanonReflect` are "proxy" types such
+as `DynamicStruct`, or any third party equivalent.
+
+## Drawbacks
+
+- Users must be aware of the difference between `CanonReflect` and `Reflect`,
+  and it must be explained.
+- You cannot directly convert a `Reflect` into a `Any` or `T`, you must first
+  make sure it is conforms a canonical representation by using either
+  `TypeRegistry::make_canon` or `<dyn Reflect>::make_canon<T: Typed>(self:
+  Box<self>)`.
+- The addition of `constructor` to `Typed` will increase the size of generated
+  code.
+- `constructor` makes `Typed` not object safe. But it should be fine
+  since `Typed` is not intended to be used as a dynamic trait objects.
+
+## Unresolved questions
+
+- `CanonReflect` name? My first instinct was `ReflectInstance`, but I changed it
+  to `ReflectType`, then `CanonicalReflect` and finally `CanonReflect` thinking
+  it was too long.
+  \
+  Technically, `Canon` doesn't mean "normal representation", as "canonical"
+  does, but it's short and close enough to be understood. I thought about
+  using `canonize` instead of `make_canon`, but it felt too religious.
+- An earlier version of this RFC had the exact opposite approach to fix the
+  uniqueness issue: Instead of removing all underlying-dependent methods from
+  `Reflect` and adding them to a new trait, we kept the underlying-dependent
+  methods but moved all the non-dependent methods to a new trait. Which one to
+  go with?
+
+## Future possibilities
+
+- Add a `target_type(&self) -> TypeId` or `full_type_name` method to `Reflect`
+  such that it's easier to check if we are converting into the proper canonical
+  type. We might also benefit from a method that is capable of taking two
+  type paths and comparing them accounting for shortcuts.
+- Performance: currently we force-convert from `CanonReflect` to `Reflect`,
+  then deeply traverse the data structure for all our reflect methods. It could
+  be greatly optimized if we add a way to tell if the underlying type is
+  already canon. (For example `make_canon` could be a no-op if the underlying
+  type is already the canonical form)
+  \
+  This also interacts interestingly with `ReflectRef` and `ReflectMut`. Should
+  the `Value` variant hold a `dyn Reflect` or a `dyn CanonReflect`?
+- `impl<T: Struct> Reflect for T {}` and other subtraits rather than
+  `Struct: Reflect`.
```

Comment:
How would this work? Wouldn't this prevent implementing/deriving `Reflect` on implementors of `Struct`?

Reply:
Yes. But with a lowered surface API (only `reflect_ref` and `reflect_mut`), there is not much point to implement `Reflect` in any specific way for implementors of `Struct`. There might be other issues, especially with `into_full`, but it's for the future.